src/sign.rs

```diff
@@ -0,0 +1,169 @@
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::convert::Infallible;
+
+use sodiumoxide::crypto::sign::ed25519;
+
+use crate::Keypair;
+#[cfg(feature = "ssh")]
+pub use ssh::*;
+
+pub trait Signer {
+    type PublicKey;
+    type Signature;
+
+    type Error;
+
+    fn public_key(&self) -> &Self::PublicKey;
+    fn sign(&mut self, data: &[u8]) -> Result<Self::Signature, Self::Error>;
+}
+
+impl Signer for Keypair<ed25519::PublicKey, ed25519::SecretKey> {
+    type PublicKey = ed25519::PublicKey;
+    type Signature = ed25519::Signature;
+
+    type Error = Infallible;
+
+    fn public_key(&self) -> &Self::PublicKey {
+        &self.public_key
+    }
+
+    fn sign(&mut self, data: &[u8]) -> Result<Self::Signature, Self::Error> {
+        Ok(ed25519::sign_detached(data, &self.secret_key))
+    }
+}
+
+#[cfg(feature = "ssh")]
+mod ssh {
+    use super::*;
+
+    use std::cell::RefCell;
+
+    pub use cryptovec::CryptoVec;
+    use futures::executor::block_on;
+    use thrussh_keys::agent::client::AgentClient;
+    use tokio::io::{AsyncRead, AsyncWrite};
+
+    pub struct SshAgent<S: AsyncRead + AsyncWrite> {
+        public_key: thrussh_keys::key::PublicKey,
+        client: RefCell<Option<AgentClient<S>>>,
+    }
+
+    impl<S: AsyncRead + AsyncWrite> SshAgent<S> {
+        pub fn new<K: Into<thrussh_keys::key::PublicKey>>(
+            public_key: K,
+            client: AgentClient<S>,
+        ) -> Self {
+            Self {
+                public_key: public_key.into(),
+                client: RefCell::new(Some(client)),
+            }
+        }
+    }
+
+    impl<S> Signer for SshAgent<S>
+    where
+        S: AsyncRead + AsyncWrite + Unpin,
+    {
+        type PublicKey = thrussh_keys::key::PublicKey;
+        type Signature = CryptoVec;
+        type Error = anyhow::Error;
+
+        fn public_key(&self) -> &Self::PublicKey {
+            &self.public_key
+        }
+
+        fn sign(&mut self, data: &[u8]) -> Result<Self::Signature, Self::Error> {
+            match self.client.replace(None) {
+                Some(client) => {
+                    self.client.replace(None);
+                    let data = CryptoVec::from(Vec::from(data));
+                    let (client, result) = block_on(client.sign_request(&self.public_key, data));
+                    let _ = self.client.replace(Some(client));
+                    result
+                },
+                None => unreachable!(),
+            }
+        }
+    }
+
+    #[cfg(test)]
+    mod tests {
```

Comment:
This is basically just the example from https://docs.rs/thrussh-keys/0.17.3/thrussh_keys, which doesn't even compile as written there. Verifying the returned signature also fails.